### ISSUE TITLE
Using deeplearning4j for face recognition (determining probability of right face)

### ISSUE THREAD

**Issue by maksgraczyk** [deeplearning4j/deeplearning4j#2615]
I'm creating a program connected with my project about face recognition. I'd like the application to:
1) Get a few images (containing a face) from a webcam.
2) Fit the model with the taken pictures (preceded by resizing them to 500x500 pixels).
3) Get another image from a webcam.
4) Determine the probability that the image (resized to 500x500 pixels before) contains the face recorded earlier.

I'm using the model described below:
```
MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(123)
                .iterations(2)
                .regularization(true).l2(0.0001)
                .learningRate(0.01)
                .weightInit(WeightInit.XAVIER)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .updater(Updater.NESTEROVS).momentum(0.9)
                .list()
                .layer(0, new ConvolutionLayer.Builder(1, 1)
                        .nIn(3)
                        .stride(1, 1)
                        .nOut(128)
                        .activation("identity")
                        .build())
                .layer(1, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(10, 10)
                        .stride(10, 10)
                        .build())
                .layer(2, new ConvolutionLayer.Builder(1, 1)
                            .stride(1, 1)
                            .nOut(200)
                            .activation("identity")
                            .build())
                .layer(3, new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(10, 10)
                        .stride(10, 10)
                        .build())
                .layer(4, new DenseLayer.Builder().activation("relu")
                        .nOut(2000).build())
                .layer(5, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .nOut(1).activation("softmax").build())
                .setInputType(InputType.convolutional(500, 500, 3))
                .backprop(true).pretrain(false)
                .build();

model = new MultiLayerNetwork(conf);
model.init();
```

I've already done the points 1, 2 and 3. The model is fitted by the following code:
```
private void fit(BufferedImage image) throws Exception
    {
        ImageIO.write(image, "png", new File("tmp.png"));
        ParentPathLabelGenerator gen = new ParentPathLabelGenerator();
        ImageRecordReader reader = new ImageRecordReader(picturePanel.getHeight(), picturePanel.getWidth(), 3, gen);
        reader.initialize(new FileSplit(new File("tmp.png")));
        DataSetIterator dataIter = new RecordReaderDataSetIterator(reader, 50, 0, 1);
        model.fit(dataIter);
    }
```
(where picturePanel.getHeight() and picturePanel.getWidth() return 500)

However, I can't manage to do the point 4. The following code:
```
private void check(BufferedImage image) throws Exception {
        ImageIO.write(image, "png", new File("tmp.png"));
        ParentPathLabelGenerator gen = new ParentPathLabelGenerator();
        ImageRecordReader reader = new ImageRecordReader(picture.getHeight(), picture.getWidth(), 3, gen);
        reader.initialize(new FileSplit(new File("tmp.png")));
        DataSetIterator dataIter = new RecordReaderDataSetIterator(reader, 50);
        while (dataIter.hasNext())
        {
            DataSet set = dataIter.next();
            INDArray array = model.output(set.getFeatures(), false);
        }
}
```
(where picture.getHeight() and picture.getWidth() return 500)
initialises the array "array" which always contains only one value: 1, regardless of the picture taken by a webcam.

I don't know what I'm doing wrong. Could you help me with the problem?

**Comment by AlexDBlack** [comment]
I'm not seeing any normalization there; that alone is enough to stop your network from learning.
Also: that network will also have a *lot* of parameters, possibly too much for your (presumably small?) data set.

Beyond that, you'll need to tune your network. Try starting with these:
https://deeplearning4j.org/troubleshootingneuralnets
https://deeplearning4j.org/visualization

As for architectures for this: you might want to look at this:
https://arxiv.org/pdf/1503.03832v3.pdf
